aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
   return linalg_norm_out_impl(result, self, c10::nullopt, ord, opt_dim, keepdim, opt_dtype);
 }
 
+Tensor linalg_tensorinv(const Tensor& self, int64_t ind) {
+  /*
+  The idea is to reduce the problem to 2D square matrix inversion.
+  Step 1. Calculate the shape of the result and the shape of the intermediate 2D matrix.
+  Step 2. Reshape `self` to 2D matrix.
+  Step 3. Invert the 2D matrix self.to_2D()
+          There is no quick way to find out whether the matrix is invertible,
+          so at this stage an error from at::inverse can be thrown.
+          Note that for CUDA this causes cross-device memory synchronization that can be slow.
+  Step 4. reshape the result.
+  */
+  TORCH_CHECK(ind > 0, "Expected a strictly positive integer for 'ind', but got ", ind);
+
+  // self[ind:]
+  std::vector<int64_t> shape_ind_end = self.sizes().slice(ind).vec();
+  // self[:ind]
+  std::vector<int64_t> shape_start_ind = self.sizes().slice(0, ind).vec();
+
+  int64_t prod_ind_end = std::accumulate(shape_ind_end.begin(), shape_ind_end.end(), int64_t{1}, std::multiplies<int64_t>());
+  int64_t prod_start_ind = std::accumulate(shape_start_ind.begin(), shape_start_ind.end(), int64_t{1}, std::multiplies<int64_t>());
+
+  // Check whether the self tensor can be reshaped to the 2D square matrix
+  TORCH_CHECK(prod_ind_end == prod_start_ind,
+    "Expected self to satisfy the requirement prod(self.shape[ind:]) == prod(self.shape[:ind]), but got ",
+    prod_ind_end, " != ", prod_start_ind);
+
+  // Concatenate shape_ind_end and shape_start_ind to form the shape of the result
+  // self[ind:] + self[:ind]
+  shape_ind_end.insert(shape_ind_end.end(), shape_start_ind.begin(), shape_start_ind.end());
+
+  // If the reshaped self is not invertible catch this error
+  Tensor result;
+  try {
+    result = at::inverse(self.reshape({prod_ind_end, prod_ind_end}));
+  } catch (...) {
+    TORCH_CHECK(false, "Failed to invert the input tensor, because it is singular.");
+  }
+
+  return result.reshape(shape_ind_end);
+}
+
+Tensor& linalg_tensorinv_out(Tensor& result, const Tensor& self, int64_t ind) {
+  CheckedFrom c = "linalg_tensorinv_out";
+  TensorArg result_arg(result, "result", 0);
+  TensorArg self_arg(result, "self", 1);
+  checkSameType(c, result_arg, self_arg);
+
+  // self[ind:]
+  std::vector<int64_t> expected_result_shape = self.sizes().slice(ind).vec();
+  // self[:ind]
+  std::vector<int64_t> shape_start_ind = self.sizes().slice(0, ind).vec();
+  // Concatenate shape_ind_end and shape_start_ind to form the shape of the result
+  // self[ind:] + self[:ind]
+  expected_result_shape.insert(expected_result_shape.end(), shape_start_ind.begin(), shape_start_ind.end());
+
+  TORCH_CHECK(result.sizes().equals(expected_result_shape),
+    "Expected result tensor to have size of ", expected_result_shape, ", but got tensor of size ", result.sizes());
+
+  Tensor result_tmp = at::linalg_tensorinv(self, ind);
+  result.copy_(result_tmp);
```

Comment:
at::inverse has an out version, can we use that to avoid the extra copy? Maybe reduce at::linalg_tensorinv without out to the one with out by passing an empty tensor, with my above comment, that should work. Only thing to be careful is that out variants do not compute gradients.

Reply:
In this case I don't think the out version of the op can be called because this is a composite op without its own autograd entry.

Reply:
Okay, I've implemented that `linalg_tensorinv_out` is primary and `linalg_tensorinv` calls to it, but autograd does not work in this case.
```
test/test_autograd.py::TestAutogradDeviceTypeCPU::test_linalg.tensorinv_cpu inverse(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.
```

Reply:
Correct. You can't use this approach. You need to keep linalg_tensorinv calling only differentiable ops (no out variants there).

Reply:
If you want to pursue this opt you could create a static helper function that takes an optional out parameter, this way you can share most of the code but add logic that if it has an out parameter you call at::inverse_out.

Reply:
linalg functions are so slow that this microoptimization of avoiding copy at the end is really not worth code complexity.